atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/ring/RingWorkflows.java

```diff
@@ -50,63 +52,85 @@ private RingWorkflows() {
-                ringWorkflowConfiguration,
-                executionExecutor);
+            AtomicBoolean skipRunning,
+            Consumer<RingValidationException> onFailure) {
+        RingWorkflowTask task = new RingWorkflowTask(skipRunning, onFailure, ringWorkflowConfiguration);
+        return DefaultWorkflow.create(store, task, ringWorkflowConfiguration, executionExecutor);
     }
 
-    private static Optional<WitnessedTransaction> run(
-            InteractiveTransactionStore store,
-            RingWorkflowConfiguration workflowConfiguration,
-            AtomicBoolean skipRunning) {
-        if (skipRunning.get()) {
-            return Optional.empty();
+    private static final class RingWorkflowTask implements KeyedTransactionTask<InteractiveTransactionStore> {
+
+        private final AtomicBoolean skipRunning;
+        private final Consumer<RingValidationException> onFailure;
+        private final RingWorkflowConfiguration workflowConfiguration;
+
+        public RingWorkflowTask(
+                AtomicBoolean skipRunning,
+                Consumer<RingValidationException> onFailure,
+                RingWorkflowConfiguration workflowConfiguration) {
+            this.skipRunning = skipRunning;
+            this.onFailure = onFailure;
+            this.workflowConfiguration = workflowConfiguration;
         }
 
-        workflowConfiguration.transactionRateLimiter().acquire();
-        String table = workflowConfiguration.tableConfiguration().tableName();
-        Integer ringSize = workflowConfiguration.ringSize();
-        return store.readWrite(txn -> {
-            Map<Integer, Optional<Integer>> data = fetchData(table, ringSize, txn);
-            try {
-                RingGraph ringGraph = RingGraph.fromPartial(data);
-                ringGraph
-                        .generateNewRing()
-                        .asMap()
-                        .forEach((rootNode, nextNode) -> txn.write(table, cell(rootNode), nextNode));
-            } catch (IllegalArgumentException e) {
-                if (skipRunning.compareAndSet(false, true)) {
-                    log.error("Detected violation with our ring.", e);
-                }
+        @Override
+        public Optional<WitnessedTransaction> apply(InteractiveTransactionStore store, Integer integer) {
+            if (skipRunning.get()) {
+                return Optional.empty();
             }
-        });
-    }
 
-    private static Map<Integer, Optional<Integer>> fetchData(
-            String table, int ringSize, InteractiveTransaction transaction) {
-        return IntStream.range(0, ringSize)
-                .boxed()
-                .collect(Collectors.toMap(Function.identity(), index -> transaction.read(table, cell(index))));
+            workflowConfiguration.transactionRateLimiter().acquire();
+            String table = workflowConfiguration.tableConfiguration().tableName();
+            Integer ringSize = workflowConfiguration.ringSize();
```

Comment:
optional/not actionable?: in hindsight it seems a bit unusual we used boxed `Integer` rather than primitive `int` for this in the ring workflow config, since we don't actually want `null` for that - but won't block on this

Reply:
switch to int, this is a bad habit!